### PR TITLE
KAFKA-13722: Refactor SerdeGetter

### PR DESCRIPTION
**Pull request by mjsax**
Refactor SerdeGetter to not use old ProcessorContext any longer.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/SerdeGetter.java [hunk 1]
```diff
@@ -18,37 +18,32 @@
...
+
+    public Serde<?> keySerde() {
+        return processorContext != null ? processorContext.keySerde() : stateStorecontext.keySerde();
     }
-    public Serde valueSerde() {
-        return oldProcessorContext != null ? oldProcessorContext.valueSerde() :
-            newProcessorContext != null ? newProcessorContext.valueSerde() : stateStorecontext.valueSerde();
+
+    public Serde<?> valueSerde() {
+        return processorContext != null ? processorContext.valueSerde() : stateStorecontext.valueSerde();
```
Comment (chia7712) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/SerdeGetter.java]: The IDE incorrectly assumes that `stateStoreContext.valueSerde()` might throw a NPE. To eliminate this false alert, perhaps we can retain the Supplier<Serde<?>> instead.
```java
    private final Supplier<Serde<?>> keySerdeSupplier;
    private final Supplier<Serde<?>> valueSerdeSupplier;

    public SerdeGetter(final ProcessorContext<?, ?> context) {
        keySerdeSupplier = context::keySerde;
        valueSerdeSupplier = context::valueSerde;
    }

    public SerdeGetter(final StateStoreContext context) {
        keySerdeSupplier = context::keySerde;
        valueSerdeSupplier = context::valueSerde;
    }

    public Serde<?> keySerde() {
        return keySerdeSupplier.get();
    }

    public Serde<?> valueSerde() {
        return valueSerdeSupplier.get();
    }
```

**Review by chia7712 — APPROVED** [review]
LGTM and one minor comment

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/SerdeGetter.java (hunk 1 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/SerdeGetter.java]: @chia7712 I like your proposal!

**Review by cadonna — APPROVED** [review]
Thanks for the PR, @mjsax !

LGTM!

**Comment by mjsax** [comment]
Java 23:
```
Found 5 test failures:
FAILED ❌ UserQuotaTest > testThrottledProducerConsumer(String, String).quorum=kraft.groupProtocol=consumer
FAILED ❌ UserQuotaTest > testQuotaOverrideDelete(String, String).quorum=kraft.groupProtocol=classic
FAILED ❌ UserQuotaTest > testQuotaOverrideDelete(String, String).quorum=kraft.groupProtocol=consumer
FAILED ❌ PlaintextAdminIntegrationTest > testConsumerGroupsDeprecatedConsumerGroupState(String, String).quorum=kraft.groupProtocol=consumer
FAILED ❌ PlaintextAdminIntegrationTest > testConsumerGroups(String, String).quorum=kraft.groupProtocol=consumer
```

**Comment by mjsax** [comment]
Java 17:
```
Found 1 test failures:
FAILED ❌ CooperativeConsumerCoordinatorTest > testOutdatedCoordinatorAssignment()
```

**Comment by chia7712** [comment]
testThrottledProducerConsumer -> https://issues.apache.org/jira/browse/KAFKA-8073
testQuotaOverrideDelete -> https://issues.apache.org/jira/browse/KAFKA-8032
testConsumerGroupsDeprecatedConsumerGroupState -> https://issues.apache.org/jira/browse/KAFKA-18298
testConsumerGroups -> https://issues.apache.org/jira/browse/KAFKA-18297
testOutdatedCoordinatorAssignment -> https://issues.apache.org/jira/browse/KAFKA-15900

**Comment by mjsax** [comment]
Cherry-picked to `4.0` branch.

**Pull request merged**